src/util/batchInsert.js

```diff
@@ -0,0 +1,84 @@
+'use strict';
+
+import {isNumber, isString, isArray, chunk, flatten} from 'lodash';
+import Promise from '../promise';
+
+export default class BatchInsert {
+	constructor(client, tableName, batch, chunkSize = 1000) {
+		if(!isNumber(chunkSize) || chunkSize < 1) {
+			throw new TypeError(`Invalid chunkSize: ${chunkSize}`);
+		}
+
+		if(!isArray(batch)) {
+			throw new TypeError(`Invalid batch: Expected array, got ${typeof batch}`)
+		}
+
+		this.client       = client;
+		this.tableName    = tableName;
+		this.batch        = chunk(batch, chunkSize);
+		this._returning   = void 0;
+		this._transaction = null;
+		this._autoCommit  = true;
+	}
+
+	/**
+	 * Columns to return from the batch operation.
+	 * @param returning
+	 */
+	returning(returning) {
+		if(isArray(returning) || isString(returning)) {
+			this._returning = returning;
+		}
+		return this;
+	}
+
+	/**
+	 * User may supply their own transaction.
+	 * If this is the case, don't autoCommit their transaction. The responsibility falls on the user.
+	 * @param transaction
+	 */
+	transacting(transaction) {
+		this._transaction = transaction;
+		this._autoCommit  = false;
+		return this;
+	}
+
+	then(callback = function() {}) {
+		let transaction;
+		return Promise.resolve()

```

Comment:
Might look more clean if you just wrap getting the transaction to separate promise and resolve it. Something like: 

```
let transactionForBatches = .... promise for getting the transaction ...;

return transactionForBatches.then(trx => {
  // here you can add the rest of promise chain and no need to store transaction to temporary variable
});
```